### ISSUE TITLE
Runtime bug: automap viewing distance buttons do nothing if held down

### ISSUE THREAD

**Issue by ghost** [dxx-rebirth/dxx-rebirth#539]
### Environment
96d108b7c907d5912a37e2ee0b54dda0d4af5ab1

#### Operating System Environment

* Microsoft Windows
  * [ ] Microsoft Windows XP
  * [ ] Microsoft Windows Vista
  * [ ] Microsoft Windows 7
  * [ ] Microsoft Windows 8
  * [ ] Microsoft Windows 8.1
  * [x] Microsoft Windows 10

* Mac OS X
  * [ ] Mac OS X 10.8
  * [ ] Mac OS X 10.9
  * [ ] Mac OS X 10.10
  * [ ] Mac OS X 10.11
  * [ ] Mac OS X 10.12
  * [ ] Mac OS X 10.13

* Debian
  * [ ] Debian Wheezy
  * [ ] Debian Jessie
  * [ ] Debian Stretch
  * [ ] Debian Buster
* Fedora
  * [ ] Fedora 26
  * [ ] Fedora 27
  * [ ] Fedora 28
  * [ ] Fedora 29
  * [ ] Fedora 30
  * [ ] Rawhide
* Ubuntu
  * [ ] Ubuntu 16.04 LTS (Xenial Xerus)
  * [ ] Ubuntu 16.10 (Yakkety Yak)
  * [ ] Ubuntu 17.04 (Zesty Zapus)
  * [ ] Ubuntu 17.10 (Artful Aardvark)
  * [ ] Ubuntu 18.04 LTS (Bionic Beaver)
  * [ ] Ubuntu 18.10 (Cosmic Cuttlefish)
  * [ ] Ubuntu 19.04 (Disco Dingo)

* [ ] Arch
* [ ] Gentoo

* [ ] OpenBSD
* [ ] FreeBSD
* [ ] NetBSD

#### CPU environment

* [ ] x86 (32-bit Intel/AMD)
* [x] x86\_64 (64-bit Intel/AMD)
* [ ] ARM (32-bit)
* [ ] ARM64 (64-bit; sometimes called AArch64)

#### Game environment
Any mission.

### Description
As part of a feature that is rarely used by most players, the automap keys F9/F10 used to control the automap viewing distance (how far the map will be shown from where the player is)  cannot be held down, like in the original game, and must instead be pressed repeatedly to adjust the range.

#### Regression status
Also present in D2X-Rebirth v0.58.x; it might be a bug as old as the adoption of SDL, and is possibly platform-dependent.

### Steps to Reproduce
1. Start a level
2. Explore the level to fill in the automap
3. Open automap
4. Hold down F9

Expected behavior: in DOS (although on a different key), after holding the button down for around a second, the range will be decreased repeatedly, similarly to how in many modern UIs, holding down a letter key will start typing the letter repeatedly after a while.

Actual behavior: holding down F9 will only shrink the range twice (once when the key is pressed down and once when it is depressed).

* Engines affected:
  * [x] D1X-Rebirth
  * [x] D2X-Rebirth

* Game modes affected:
  * [x] Single player
  * [ ] Multiplayer competitive
    * [ ] Anarchy
    * [ ] Team Anarchy
    * [ ] Robo-Anarchy
    * [ ] Capture the Flag
    * [ ] Bounty
  * [ ] Multiplayer cooperative

**Comment by vLKp** [comment]
Now that you filed this, I remember seeing it a long time ago (v0.57.3, I think).  I ended up not using the feature because of it, and by the time I took up maintaining Rebirth, I had forgotten the problem.

**Comment by InfernalCore** [comment]
Hello Maintainers. 
I noticed this issue with F9/F10 in d1x-d2x, releases c49598d and 5c900ca, compiled for Linux (Mint 21.0 Mint 21.2) on a 64-bit Intel platform.
I find the F9/F10 feature useful, even if it doesn't seem very popular among the players. 
I solved this issue as follows. This solution is easy and it seems to work without creating other problems in D1/D2 (both the above mentioned releases). I post it here so others may use it. Feel free to incorporate it in the code, if you find it acceptable. 
Caveat: I am not a real programmer.
Congratulations and Keep up the great work!
Infernal Core.

FILE: automap.cpp
Line:  1139

    switch (event.type)
    {
        case EVENT_WINDOW_ACTIVATED:
            game_flush_inputs(controls);
            event_toggle_focus(1);
            key_toggle_repeat(1);  // MODIFIED. IT WAS  key_toggle_repeat(0);
            break;

        case EVENT_WINDOW_DEACTIVATED:
            event_toggle_focus(0);
            key_toggle_repeat(1);
            break;

**Comment by vLKp** [comment]
As you suspect, it has broader impact, since all key repetition is enabled.  It seems like a plausible fix, but I would like to work through what other keys become repeatable and verify none of them cause problems.  What tests have you tried with it to date?

**Comment by InfernalCore** [comment]
Thanks for your prompt acknowledgement. 
As you say, I suspect that my hack may have a broader impact, however I wasn't able to find anything yet. 

I did not carry out any specific test, but I will check out any list you may suggest. 

I play SP only (no MP). I don't play much D1.
Actually I am just doing the things I use to do when playing D2 (SP) such as: 
- create new games
- open savegames
- save games
- move 6 dof (JS+KB)
- switch weapons (KB+JS)
- turn light on/off (JS, key H)
- toggle rear view (short press R key)
- toggle rear view (long press R key)
- shot (JS, mouse)
- record demo
- play demo (stop/play/fwd/rev)
- open map (navigate map)
- gather power-ups
- destroy bots
- be destroyed by bots

I can only guess that key_toggle_repeat(1) affects only the Map environment since it is placed in automap.cpp, but not being a coder myself, it 's hard to tell. In any case I will go on playing as usual; as soon as I find out something weird I will post a note here. 

Regards!

**Comment by InfernalCore** [comment]
I forgot to mention. I tried also: 
- drop bomb (key C for me)
- switch bomb prox/smart (key B for me)
- Rotate primary/secondary weapons with keys CAPSLOCK and SPACE BAR.
- E-S Converter (key T)
The latter (T) has key repetition, but it works the same way also playing unaltered D2X-Rebirth. So I think this should be the expected behavior. 
I am quite sure I played DESCENT2: COUNTERSTRIKE all levels 1 to 7, plus all Brimspark and Omega with the hacked version. Not sure about the other levels. 

I hope others will read this post and try themselves. The hack seems to impact only the Map environment to me. Still some bad behavior may lurk unseen. If you have any specific test to run don't hesitate to ask (such as some configuration or setting or more esoteric things I'm not aware of) . 

Regards.

**Comment by vLKp** [comment]
I agree, this ought to impact only the automap.  This is not specifically because the modified code is in `automap.cpp`, but rather because deactivating the automap already makes the call you propose to make in the activation path.  Therefore, trying all the various automap functions is likely sufficient testing.  Any non-automap activity would already be run with key repeat enabled even without your change.

**Comment by InfernalCore** [comment]
Understood.

As for the Automap environment:
- Navigation works as expected (with option "free navigation" enabled or not).
- In D2, selecting and clearing Markers (#, CTRL-D) works as expected.
- Nothing strange happens when keeping keys pressed.
- There's not much to test here AFAIK.

During the last month I played the whole "D1:First Strike" with the (hacked) commit 5c900ca in Linux Mint 21.2 Cinnamon, kernel 5.15.0, x86_64.

I'm not so familiar with D1, so I kept calling the Automap practically at every door.
F9/F10 are useful when the player is lost in a jam of locally intersecting tunnels: just keeping F9 pressed clears out the overlapping lines.

In all this time, nothing weird came up.

5c900ca is getting old, time for me to build a new one.

Best regards.
Infernal Core.
